common/core/gpudevice.cpp

```diff
@@ -41,126 +41,171 @@ bool GpuDevice::Initialize() {
+  std::vector<bool> available_displays(temp_displays.size(), true);
 
   if (use_mosiac) {
-    int32_t l_size = physical_displays.size();
-    mdisplays.emplace_back();
-    std::vector<NativeDisplay *> &temp = mdisplays.back();
-    for (int32_t i = 0; i < l_size; i++) {
-      hwcomposer::NativeDisplay *display = physical_displays.at(i);
-      temp.emplace_back(display);
-
-      if (least_mosiac_index == -1) {
-        least_mosiac_index = i;
-      } else {
-        least_mosiac_index = std::min(least_mosiac_index, i);
+    for (size_t i = 0; i < mosiac_displays.size(); i++) {
+      std::vector<NativeDisplay *> i_available_mosiac_displays;
+      for (size_t j = 0; j < mosiac_displays.at(i).size(); j++) {
+        if (mosiac_displays.at(i).at(j) >= 0 &&
+            mosiac_displays.at(i).at(j) < available_displays.size()) {
+          i_available_mosiac_displays.emplace_back(
+              temp_displays.at(mosiac_displays.at(i).at(j)));
+          // Add tag for no mosiac-ed displays
+          available_displays.at(mosiac_displays.at(i).at(j)) = false;
+        }
       }
+      std::unique_ptr<MosiacDisplay> mosiac(
+          new MosiacDisplay(i_available_mosiac_displays));
+      mosiac_displays_.emplace_back(std::move(mosiac));
     }
 
-    std::unique_ptr<MosiacDisplay> mosiac(new MosiacDisplay(temp));
-    mosiac_displays_.emplace_back(std::move(mosiac));
-  }
-
-  if (least_mosiac_index == 0) {
-    total_displays_.emplace_back(mosiac_displays_.at(0).get());
-  } else {
-    if (least_mosiac_index == -1) {
-      least_mosiac_index = physical_displays.size();
-    }
-
-    for (int32_t i = 0; i < least_mosiac_index; i++) {
-      total_displays_.emplace_back(physical_displays.at(i));
+    // Save mosiac displays to final total_displays_
+    for (size_t i = 0; i < mosiac_displays_.size(); i++) {
+      total_displays_.emplace_back(mosiac_displays_.at(i).get());
     }
   }
 
-  if (least_mosiac_index == static_cast<int32_t>(physical_displays.size())) {
-    return true;
-  }
-
-  uint32_t m_size = mdisplays.size();
-  uint32_t mosiac_index_added = least_mosiac_index;
-  least_mosiac_index++;
-  for (size_t i = least_mosiac_index; i < size; i++) {
-    hwcomposer::NativeDisplay *display = physical_displays.at(i);
-    bool skip_display = false;
-    for (size_t i = 0; i < m_size; ++i) {
-      std::vector<NativeDisplay *> &temp_display = mdisplays.at(i);
-      size_t temp = temp_display.size();
-      for (size_t i = 0; i < temp; ++i) {
-        if (temp_display.at(i) == display) {
-          skip_display = true;
-          if (mosiac_index_added < i) {
-            total_displays_.emplace_back(mosiac_displays_.at(0).get());
-            mosiac_index_added = i;
-          }
-
-          break;
-        }
-      }
+  // Save save no mosiac displays to final total_displays_
+  for (size_t i = 0; i < available_displays.size(); i++) {
+    if (available_displays.at(i)) {
+      total_displays_.emplace_back(temp_displays.at(i));
```

Comment:
these should be sorted by order. i.e. Mosiac Display having display 1 should be first in list and so on. It could be that Logical display could have 1 and be part of Mosiac display etc.

Reply:
or could be that physical display 1 is just an independent display